File: src/xyz/human_resource_machine/interpreter.py
import logging
import typing
from dataclasses import dataclass

logger = logging.getLogger(__name__)

Value = typing.Union[int, str]


def int_or_str(value: str) -> Value:
    """Convert a string to an int if possible, otherwise return the string."""
    try:
        return int(value)
    except ValueError:
        return value


@dataclass(frozen=True, slots=True)
class _UsesRegister:
    register: Value
    indirect: bool = False


@dataclass(frozen=True, slots=True)
class Comment:
    text: str


@dataclass(frozen=True, slots=True)
class Inbox:
    pass


@dataclass(frozen=True, slots=True)
class Outbox:
    pass


@dataclass(frozen=True, slots=True)
class Add(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class Subtract(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class CopyTo(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class CopyFrom(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class BumpPlus(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class BumpMinus(_UsesRegister):
    pass


@dataclass(frozen=True, slots=True)
class Label:
    label: str


@dataclass(frozen=True, slots=True)
class Jump:
    label: str


@dataclass(frozen=True, slots=True)
class JumpIfZero:
    label: str


@dataclass(frozen=True, slots=True)
class JumpIfNegative:
    label: str


@dataclass(frozen=True, slots=True)
class AssertValueIs:
    """This instruction does not exist in the original game, but is useful for testing."""

    value: Value


@dataclass(frozen=True, slots=True)
class AssertRegisterIs:
    """This instruction does not exist in the original game, but is useful for testing."""

    register: str
    value: Value


Instruction = typing.Union[
    Comment,
    Inbox,
    Outbox,
    Add,
    Subtract,
    CopyTo,
    CopyFrom,
    BumpPlus,
    BumpMinus,
    Label,
    Jump,
    JumpIfZero,
    JumpIfNegative,
    AssertValueIs,
]


class Interpreter:
    def __init__(
        self,
        *,
        registers: dict[Value, Value] | None = None,
        instructions: list[Instruction] | None = None,
        input: list[Value] | None = None,
    ):
        self.instructions = [] if instructions is None else instructions.copy()
        self.registers = {} if registers is None else registers.copy()
        self.instruction_count: int = 0
        self._value: int | None = None
        self._input = [] if input is None else input
        self._input_index: int = 0
        self._execution_count: int = 0
        self._instruction_index: int = 0
        self._output: list[Value] = []

        self.instruction_count = 0
        for instruction in self.instructions:
            match instruction:
                case Label() | Comment():
                    pass
                case _:
                    self.instruction_count += 1

        self._jumps: dict[str, int] = {}
        for index, instruction in enumerate(self.instructions):
            match instruction:
                case Label():
                    self._jumps[instruction.label] = index
                case _:
                    pass

    def _read_register(self, instruction: _UsesRegister) -> Value:
        """Read the value from the register, handling indirect addressing."""
        if instruction.indirect:
            return self.registers[self.registers[instruction.register]]
        return self.registers[instruction.register]

    def _write_register(self, instruction: _UsesRegister) -> None:
        """Write the value to the register, handling indirect addressing."""
        if instruction.indirect:
            self.registers[self.registers[instruction.register]] = self._value
        else:
            self.registers[instruction.register] = self._value

    def execute_program(self) -> list[Value]:
        """Execute all the instructions in the program until completion."""
        while self._instruction_index < len(self.instructions):
            return_value = self.step()
            if return_value is not None:
                return return_value
        return self.output

    def step(self) -> list[Value] | None:
        """Execute the next instruction in the program."""
        instruction = self.instructions[self._instruction_index]
        logger.debug("Instruction %d: %s", self._instruction_index, instruction)
        match instruction:
            case Inbox():
                if self._input_index >= len(self._input):
                    return self.output  # No more input available.
                self._value = self._input[self._input_index]
                self._input_index += 1
                self._instruction_index += 1
                self._execution_count += 1
            case Outbox():
                if self._value is None:
                    raise ValueError("No value to output")
                self._output.append(self._value)
                self._value = None
                self._instruction_index += 1
                self._execution_count += 1
            case Add() as add_:
                if not isinstance(self._value, int):
                    raise ValueError(
                        f"Value {self._value} must be an integer for addition"
                    )
                argument = self._read_register(add_)
                if not isinstance(argument, int):
                    raise ValueError(
                        f"Argument {argument} must be an integer for addition"
                    )
                self._value += argument
                self._instruction_index += 1
                self._execution_count += 1
            case Subtract() as subtract_:
                if not isinstance(self._value, int):
                    raise ValueError(
                        f"Value {self._value} must be an integer for subtraction"
                    )
                argument = self._read_register(subtract_)
                if not isinstance(argument, int):
                    raise ValueError(
                        f"Argument {argument} must be an integer for subtraction"
                    )
                self._value -= argument
                self._instruction_index += 1
                self._execution_count += 1
            case CopyTo() as copy_to:
                self._write_register(copy_to)
                self._instruction_index += 1
                self._execution_count += 1
            case CopyFrom() as copy_from:
                register_value = self._read_register(copy_from)
                self._value = register_value
                self._instruction_index += 1
                self._execution_count += 1
            case BumpPlus() as bump_plus:
                if bump_plus.indirect:
                    self.registers[self.registers[bump_plus.register]] += 1
                    self._value = self.registers[self.registers[bump_plus.register]]
                else:
                    self.registers[bump_plus.register] += 1
                    self._value = self.registers[bump_plus.register]
                self._instruction_index += 1
                self._execution_count += 1
            case BumpMinus() as bump_minus:
                if bump_minus.indirect:
                    self.registers[self.registers[bump_minus.register]] -= 1
                    self._value = self.registers[self.registers[bump_minus.register]]
                else:
                    self.registers[bump_minus.register] -= 1
                    self._value = self.registers[bump_minus.register]
                self._instruction_index += 1
                self._execution_count += 1
            case Label() | Comment():
                self._instruction_index += 1
                # No execution count increment for comments or jump targets.
            case Jump() as jump:
                self._instruction_index = self._jumps[jump.label]
                self._execution_count += 1
            case JumpIfZero() as jump:
                if self._value == 0:
                    self._instruction_index = self._jumps[jump.label]
                else:
                    self._instruction_index += 1
                self._execution_count += 1
            case JumpIfNegative() as jump:
                if self._value < 0:
                    self._instruction_index = self._jumps[jump.label]
                else:
                    self._instruction_index += 1
                self._execution_count += 1
            case AssertValueIs() as assert_value:
                if self._value != assert_value.value:
                    raise ValueError(
                        f"Assertion failed: expected {assert_value.value}, got {self._value}"
                    )
                self._instruction_index += 1
            case AssertRegisterIs() as assert_register:
                if self.registers[assert_register.register] != assert_register.value:
                    raise ValueError(
                        f"Assertion failed: expected register '{assert_register.register}' to be {assert_register.value}, got {self.registers[assert_register.register]}"
                    )
                self._instruction_index += 1

    def to_str(self) -> str:
        lines: list[str] = []
        index: int = 1

        for instruction in self.instructions:
            match instruction:
                case AssertValueIs() | AssertRegisterIs():
                    continue
                case Comment() as comment:
                    lines.append(f'"{comment.text}"')
                case Label() as label:
                    lines.append(f"LABEL: {label.label}")
                case _ as instruction:
                    lines.append(f"{index}: {instruction}")
                    index += 1
        return "\n".join(lines)

    def to_dot(self) -> str:
        """Return the program instructions in dotfile format."""
        lines: list[str] = ["digraph G {"]
        node_map: dict[int, str] = {}
        # First pass: define nodes and labels for actual instructions
        idx_counter = 0
        for i, instruction in enumerate(self.instructions):
            match instruction:
                case Comment():
                    # Comments are not nodes in the graph
                    pass
                case Label() as label_instr:
                    # Labels point to the next actual instruction
                    node_map[i] = f'"{label_instr.label}"'
                    lines.append(f"  {node_map[i]} [shape=plaintext];")
                case _:
                    node_id = f"instr_{idx_counter}"
                    node_map[i] = node_id
                    # Sanitize instruction string for dot format
                    instr_str = str(instruction).replace(
                        '"', '\\"'
                    )  # Corrected sanitization
                    shape = "box"
                    if isinstance(instruction, (Jump, JumpIfZero, JumpIfNegative)):
                        shape = "diamond"
                    lines.append(
                        f'  {node_id} [shape={shape}, label="{idx_counter}: {instr_str}"];'
                    )
                    idx_counter += 1

        # Second pass: define edges
        for i, instruction in enumerate(self.instructions):
            if i not in node_map:  # Skip comments
                continue

            current_node_id = node_map[i]

            # Determine the actual next instruction index, skipping comments
            actual_next_instr_idx = i + 1
            while actual_next_instr_idx < len(self.instructions) and isinstance(
                self.instructions[actual_next_instr_idx], Comment
            ):
                actual_next_instr_idx += 1

            # Default sequential flow (if not a jump or label that only points elsewhere)
            if not isinstance(instruction, (Jump, JumpIfZero, JumpIfNegative, Label)):
                if (
                    actual_next_instr_idx < len(self.instructions)
                    and actual_next_instr_idx in node_map
                ):
                    lines.append(
                        f"  {current_node_id} -> {node_map[actual_next_instr_idx]};"
                    )
            elif isinstance(
                instruction, Label
            ):  # Labels should point to the next instruction
                if (
                    actual_next_instr_idx < len(self.instructions)
                    and actual_next_instr_idx in node_map
                ):
                    lines.append(
                        f"  {current_node_id} -> {node_map[actual_next_instr_idx]};"
                    )

            match instruction:
                case Jump() as jump_instr:
                    target_idx = self._jumps.get(jump_instr.label)
                    if target_idx is not None and target_idx in node_map:
                        lines.append(f"  {current_node_id} -> {node_map[target_idx]};")
                case JumpIfZero() as jump_if_zero_instr:
                    target_idx = self._jumps.get(jump_if_zero_instr.label)
                    if target_idx is not None and target_idx in node_map:
                        lines.append(
                            f'  {current_node_id} -> {node_map[target_idx]} [label="if zero"];'
                        )
                    # Edge for not jumping (sequential)
                    if (
                        actual_next_instr_idx < len(self.instructions)
                        and actual_next_instr_idx in node_map
                    ):
                        lines.append(
                            f'  {current_node_id} -> {node_map[actual_next_instr_idx]} [label="not zero"];'
                        )
                case JumpIfNegative() as jump_if_negative_instr:
                    target_idx = self._jumps.get(jump_if_negative_instr.label)
                    if target_idx is not None and target_idx in node_map:
                        lines.append(
                            f'  {current_node_id} -> {node_map[target_idx]} [label="if negative"];'
                        )
                    # Edge for not jumping (sequential)
                    if (
                        actual_next_instr_idx < len(self.instructions)
                        and actual_next_instr_idx in node_map
                    ):
                        lines.append(
                            f'  {current_node_id} -> {node_map[actual_next_instr_idx]} [label="not negative"];'
                        )
                # For other instructions like Inbox, Outbox, Add, Subtract, CopyTo, CopyFrom, BumpPlus, BumpMinus,
                # the default sequential link is already added above if they are not jumps or labels.
                # No special edge logic needed here for them beyond the default.
                case _:
                    pass

        lines.append("}")
        return "\n".join(lines)

    @property
    def value(self) -> Value | None:
        return self._value

    def register(self, value: Value) -> Value | None:
        return self.registers.get(value)

    @property
    def output(self) -> list[Value]:
        return self._output.copy()

    @property
    def executions(self) -> int:
        return self._execution_count

    @property
    def instruction_index(self) -> int:
        return self._instruction_index

File: rust/src/interpreter.rs
use log::debug;
use std::collections::HashMap;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Value {
    Int(i32),
    String(String),
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Value::Int(i) => write!(f, "{}", i),
            Value::String(s) => write!(f, "{}", s),
        }
    }
}

impl Value {
    pub fn as_int(&self) -> Result<i32, String> {
        match self {
            Value::Int(i) => Ok(*i),
            Value::String(_) => Err(format!("Expected integer, got string: {}", self)),
        }
    }
}

#[derive(Debug, Clone)]
pub struct UsesRegister {
    pub register: Value,
    pub indirect: bool,
}

impl UsesRegister {
    pub fn new(register: Value) -> Self {
        Self {
            register,
            indirect: false,
        }
    }

    pub fn new_indirect(register: Value) -> Self {
        Self {
            register,
            indirect: true,
        }
    }
}

#[derive(Debug, Clone)]
pub enum Instruction {
    Comment { text: String },
    Inbox,
    Outbox,
    Add(UsesRegister),
    Subtract(UsesRegister),
    CopyTo(UsesRegister),
    CopyFrom(UsesRegister),
    BumpPlus(UsesRegister),
    BumpMinus(UsesRegister),
    Label { label: String },
    Jump { label: String },
    JumpIfZero { label: String },
    JumpIfNegative { label: String },
    AssertValueIs { value: Value },
    AssertRegisterIs { register: String, value: Value },
}

impl fmt::Display for Instruction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Instruction::Comment { text } => write!(f, "# {}", text),
            Instruction::Inbox => write!(f, "INBOX"),
            Instruction::Outbox => write!(f, "OUTBOX"),
            Instruction::Add(reg) => {
                if reg.indirect {
                    write!(f, "ADD [{}]", reg.register)
                } else {
                    write!(f, "ADD {}", reg.register)
                }
            }
            Instruction::Subtract(reg) => {
                if reg.indirect {
                    write!(f, "SUB [{}]", reg.register)
                } else {
                    write!(f, "SUB {}", reg.register)
                }
            }
            Instruction::CopyTo(reg) => {
                if reg.indirect {
                    write!(f, "COPYTO [{}]", reg.register)
                } else {
                    write!(f, "COPYTO {}", reg.register)
                }
            }
            Instruction::CopyFrom(reg) => {
                if reg.indirect {
                    write!(f, "COPYFROM [{}]", reg.register)
                } else {
                    write!(f, "COPYFROM {}", reg.register)
                }
            }
            Instruction::BumpPlus(reg) => {
                if reg.indirect {
                    write!(f, "BUMPUP [{}]", reg.register)
                } else {
                    write!(f, "BUMPUP {}", reg.register)
                }
            }
            Instruction::BumpMinus(reg) => {
                if reg.indirect {
                    write!(f, "BUMPDN [{}]", reg.register)
                } else {
                    write!(f, "BUMPDN {}", reg.register)
                }
            }
            Instruction::Label { label } => write!(f, "LABEL: {}", label),
            Instruction::Jump { label } => write!(f, "JUMP {}", label),
            Instruction::JumpIfZero { label } => write!(f, "JUMPZ {}", label),
            Instruction::JumpIfNegative { label } => write!(f, "JUMPN {}", label),
            Instruction::AssertValueIs { value } => write!(f, "ASSERT_VALUE {}", value),
            Instruction::AssertRegisterIs { register, value } => {
                write!(f, "ASSERT_REGISTER {} {}", register, value)
            }
        }
    }
}

pub struct Interpreter {
    instructions: Vec<Instruction>,
    registers: HashMap<Value, Value>,
    instruction_count: usize,
    value: Option<Value>,
    input: Vec<Value>,
    input_index: usize,
    execution_count: usize,
    instruction_index: usize,
    output: Vec<Value>,
    jumps: HashMap<String, usize>,
}

impl Interpreter {
    pub fn new(
        registers: HashMap<Value, Value>,
        instructions: Vec<Instruction>,
        input: Vec<Value>,
    ) -> Self {
        let mut interpreter = Self {
            instructions: instructions.clone(),
            registers,
            instruction_count: 0,
            value: None,
            input,
            input_index: 0,
            execution_count: 0,
            instruction_index: 0,
            output: Vec::new(),
            jumps: HashMap::new(),
        };

        // Count actual instructions (excluding labels and comments)
        interpreter.instruction_count = instructions
            .iter()
            .filter(|instr| {
                !matches!(
                    instr,
                    Instruction::Label { .. } | Instruction::Comment { .. }
                )
            })
            .count();

        // Build jump table
        for (index, instruction) in instructions.iter().enumerate() {
            if let Instruction::Label { label } = instruction {
                interpreter.jumps.insert(label.clone(), index);
            }
        }

        interpreter
    }

    fn read_register(&self, uses_register: &UsesRegister) -> Result<Value, String> {
        if uses_register.indirect {
            let indirect_key = self
                .registers
                .get(&uses_register.register)
                .ok_or_else(|| format!("Register {:?} not found", uses_register.register))?;
            self.registers
                .get(indirect_key)
                .cloned()
                .ok_or_else(|| format!("Indirect register {:?} not found", indirect_key))
        } else {
            self.registers
                .get(&uses_register.register)
                .cloned()
                .ok_or_else(|| format!("Register {:?} not found", uses_register.register))
        }
    }

    fn write_register(&mut self, uses_register: &UsesRegister) -> Result<(), String> {
        let value = self
            .value
            .clone()
            .ok_or_else(|| "No value to write to register".to_string())?;

        if uses_register.indirect {
            let indirect_key = self
                .registers
                .get(&uses_register.register)
                .cloned()
                .ok_or_else(|| format!("Register {:?} not found", uses_register.register))?;
            self.registers.insert(indirect_key, value);
        } else {
            self.registers.insert(uses_register.register.clone(), value);
        }
        Ok(())
    }

    pub fn execute_program(&mut self) -> Result<Vec<Value>, String> {
        while self.instruction_index < self.instructions.len() {
            if let Some(output) = self.step()? {
                return Ok(output);
            }
        }
        Ok(self.output.clone())
    }

    pub fn step(&mut self) -> Result<Option<Vec<Value>>, String> {
        if self.instruction_index >= self.instructions.len() {
            return Ok(Some(self.output.clone()));
        }

        let instruction = self.instructions[self.instruction_index].clone();
        debug!("Instruction {}: {:?}", self.instruction_index, instruction);

        match instruction {
            Instruction::Inbox => {
                if self.input_index >= self.input.len() {
                    return Ok(Some(self.output.clone())); // No more input available
                }
                self.value = Some(self.input[self.input_index].clone());
                self.input_index += 1;
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::Outbox => {
                let value = self
                    .value
                    .take()
                    .ok_or_else(|| "No value to output".to_string())?;
                self.output.push(value);
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::Add(uses_register) => {
                let current_value = self
                    .value
                    .as_ref()
                    .ok_or_else(|| "No value for addition".to_string())?
                    .as_int()?;
                let argument = self.read_register(&uses_register)?.as_int()?;
                self.value = Some(Value::Int(current_value + argument));
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::Subtract(uses_register) => {
                let current_value = self
                    .value
                    .as_ref()
                    .ok_or_else(|| "No value for subtraction".to_string())?
                    .as_int()?;
                let argument = self.read_register(&uses_register)?.as_int()?;
                self.value = Some(Value::Int(current_value - argument));
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::CopyTo(uses_register) => {
                self.write_register(&uses_register)?;
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::CopyFrom(uses_register) => {
                self.value = Some(self.read_register(&uses_register)?);
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::BumpPlus(uses_register) => {
                if uses_register.indirect {
                    let indirect_key = self
                        .registers
                        .get(&uses_register.register)
                        .cloned()
                        .ok_or_else(|| {
                            format!("Register {:?} not found", uses_register.register)
                        })?;
                    let current = self
                        .registers
                        .get(&indirect_key)
                        .ok_or_else(|| format!("Indirect register {:?} not found", indirect_key))?
                        .as_int()?;
                    let new_value = Value::Int(current + 1);
                    self.registers.insert(indirect_key, new_value.clone());
                    self.value = Some(new_value);
                } else {
                    let current = self
                        .registers
                        .get(&uses_register.register)
                        .ok_or_else(|| format!("Register {:?} not found", uses_register.register))?
                        .as_int()?;
                    let new_value = Value::Int(current + 1);
                    self.registers
                        .insert(uses_register.register.clone(), new_value.clone());
                    self.value = Some(new_value);
                }
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::BumpMinus(uses_register) => {
                if uses_register.indirect {
                    let indirect_key = self
                        .registers
                        .get(&uses_register.register)
                        .cloned()
                        .ok_or_else(|| {
                            format!("Register {:?} not found", uses_register.register)
                        })?;
                    let current = self
                        .registers
                        .get(&indirect_key)
                        .ok_or_else(|| format!("Indirect register {:?} not found", indirect_key))?
                        .as_int()?;
                    let new_value = Value::Int(current - 1);
                    self.registers.insert(indirect_key, new_value.clone());
                    self.value = Some(new_value);
                } else {
                    let current = self
                        .registers
                        .get(&uses_register.register)
                        .ok_or_else(|| format!("Register {:?} not found", uses_register.register))?
                        .as_int()?;
                    let new_value = Value::Int(current - 1);
                    self.registers
                        .insert(uses_register.register.clone(), new_value.clone());
                    self.value = Some(new_value);
                }
                self.instruction_index += 1;
                self.execution_count += 1;
            }
            Instruction::Label { .. } | Instruction::Comment { .. } => {
                self.instruction_index += 1;
                // No execution count increment for comments or labels
            }
            Instruction::Jump { label } => {
                self.instruction_index = *self
                    .jumps
                    .get(&label)
                    .ok_or_else(|| format!("Label '{}' not found", label))?;
                self.execution_count += 1;
            }
            Instruction::JumpIfZero { label } => {
                let should_jump = match &self.value {
                    Some(Value::Int(0)) => true,
                    Some(Value::String(s)) if s == "0" => true,
                    _ => false,
                };

                if should_jump {
                    self.instruction_index = *self
                        .jumps
                        .get(&label)
                        .ok_or_else(|| format!("Label '{}' not found", label))?;
                } else {
                    self.instruction_index += 1;
                }
                self.execution_count += 1;
            }
            Instruction::JumpIfNegative { label } => {
                let should_jump = match &self.value {
                    Some(Value::Int(i)) if *i < 0 => true,
                    Some(Value::String(s)) => s.parse::<i32>().map(|i| i < 0).unwrap_or(false),
                    _ => false,
                };

                if should_jump {
                    self.instruction_index = *self
                        .jumps
                        .get(&label)
                        .ok_or_else(|| format!("Label '{}' not found", label))?;
                } else {
                    self.instruction_index += 1;
                }
                self.execution_count += 1;
            }
            Instruction::AssertValueIs { value } => {
                if self.value.as_ref() != Some(&value) {
                    return Err(format!(
                        "Assertion failed: expected {:?}, got {:?}",
                        value, self.value
                    ));
                }
                self.instruction_index += 1;
            }
            Instruction::AssertRegisterIs { register, value } => {
                let register_key = Value::String(register);
                let actual = self.registers.get(&register_key);
                if actual != Some(&value) {
                    return Err(format!(
                        "Assertion failed: expected register '{}' to be {:?}, got {:?}",
                        register_key, value, actual
                    ));
                }
                self.instruction_index += 1;
            }
        }

        Ok(None)
    }

    pub fn to_string(&self) -> String {
        let mut lines = Vec::new();
        let mut index = 1;

        for instruction in &self.instructions {
            match instruction {
                Instruction::AssertValueIs { .. } | Instruction::AssertRegisterIs { .. } => {
                    continue
                }
                Instruction::Comment { text } => {
                    lines.push(format!("\"{}\"", text));
                }
                Instruction::Label { label } => {
                    lines.push(format!("LABEL: {}", label));
                }
                _ => {
                    lines.push(format!("{}: {}", index, instruction));
                    index += 1;
                }
            }
        }

        lines.join("\n")
    }

    pub fn to_dot(&self) -> String {
        let mut lines = vec!["digraph G {".to_string()];
        let mut node_map = HashMap::new();
        let mut idx_counter = 0;

        // First pass: define nodes and labels for actual instructions
        for (i, instruction) in self.instructions.iter().enumerate() {
            match instruction {
                Instruction::Comment { .. } => {
                    // Comments are not nodes in the graph
                }
                Instruction::Label { label } => {
                    let node_id = format!("\"{}\"", label);
                    node_map.insert(i, node_id.clone());
                    lines.push(format!("  {} [shape=plaintext];", node_id));
                }
                _ => {
                    let node_id = format!("instr_{}", idx_counter);
                    node_map.insert(i, node_id.clone());
                    let instr_str = instruction.to_string();
                    let shape = match instruction {
                        Instruction::Jump { .. }
                        | Instruction::JumpIfZero { .. }
                        | Instruction::JumpIfNegative { .. } => "diamond",
                        _ => "box",
                    };
                    lines.push(format!(
                        "  {} [shape={}, label=\"{}: {}\"];",
                        node_id, shape, idx_counter, instr_str
                    ));
                    idx_counter += 1;
                }
            }
        }

        // Second pass: define edges
        for (i, instruction) in self.instructions.iter().enumerate() {
            if !node_map.contains_key(&i) {
                continue; // Skip comments
            }

            let current_node_id = &node_map[&i];

            // Find next instruction index, skipping comments
            let mut next_instr_idx = i + 1;
            while next_instr_idx < self.instructions.len()
                && matches!(
                    self.instructions[next_instr_idx],
                    Instruction::Comment { .. }
                )
            {
                next_instr_idx += 1;
            }

            let next_node_exists =
                next_instr_idx < self.instructions.len() && node_map.contains_key(&next_instr_idx);
            let next_node_id = node_map.get(&next_instr_idx);

            match instruction {
                Instruction::Jump { label } => {
                    if let Some(target_idx) = self.jumps.get(label) {
                        if let Some(target_node) = node_map.get(target_idx) {
                            lines.push(format!("  {} -> {};", current_node_id, target_node));
                        }
                    }
                }
                Instruction::JumpIfZero { label } => {
                    if let Some(target_idx) = self.jumps.get(label) {
                        if let Some(target_node) = node_map.get(target_idx) {
                            lines.push(format!(
                                "  {} -> {} [label=\"if zero\"];",
                                current_node_id, target_node
                            ));
                        }
                    }
                    if next_node_exists {
                        if let Some(next_node) = next_node_id {
                            lines.push(format!(
                                "  {} -> {} [label=\"not zero\"];",
                                current_node_id, next_node
                            ));
                        }
                    }
                }
                Instruction::JumpIfNegative { label } => {
                    if let Some(target_idx) = self.jumps.get(label) {
                        if let Some(target_node) = node_map.get(target_idx) {
                            lines.push(format!(
                                "  {} -> {} [label=\"if negative\"];",
                                current_node_id, target_node
                            ));
                        }
                    }
                    if next_node_exists {
                        if let Some(next_node) = next_node_id {
                            lines.push(format!(
                                "  {} -> {} [label=\"not negative\"];",
                                current_node_id, next_node
                            ));
                        }
                    }
                }
                Instruction::Label { .. } => {
                    // Labels point to the next actual instruction
                    if next_node_exists {
                        if let Some(next_node) = next_node_id {
                            lines.push(format!("  {} -> {};", current_node_id, next_node));
                        }
                    }
                }
                Instruction::Comment { .. } => {
                    // Comments are already skipped
                }
                _ => {
                    // Default sequential flow for other instructions
                    if next_node_exists {
                        if let Some(next_node) = next_node_id {
                            lines.push(format!("  {} -> {};", current_node_id, next_node));
                        }
                    }
                }
            }
        }

        lines.push("}".to_string());
        lines.join("\n")
    }

    pub fn executions(&self) -> usize {
        self.execution_count
    }

    pub fn instruction_count(&self) -> usize {
        self.instruction_count
    }

    pub fn registers_len(&self) -> usize {
        self.registers.len()
    }

    pub fn input_string(&self) -> String {
        self.input
            .iter()
            .map(|v| v.to_string())
            .collect::<Vec<_>>()
            .join(", ")
    }
}
